internal/dataplane/kongstate/kongstate.go

```diff
@@ -196,17 +204,46 @@ func (ks *KongState) FillOverrides(logger logr.Logger, s store.Storer) {
 		}
 	}
 
-	// Upstreams
+	ks.FillUpstreamOverrides(s, logger, failuresCollector)
+}
+
+func (ks *KongState) FillUpstreamOverrides(
+	s store.Storer,
+	logger logr.Logger,
+	failuresCollector *failures.ResourceFailuresCollector,
+) {
 	for i := 0; i < len(ks.Upstreams); i++ {
-		kongIngress, err := getKongIngressForServices(s, ks.Upstreams[i].Service.K8sServices)
+		servicesGroup := lo.Values(ks.Upstreams[i].Service.K8sServices)
+		servicesAsObjects := func(svc *corev1.Service, _ int) client.Object { return svc }
+
+		kongIngress, err := getKongIngressForServices(s, servicesGroup)
 		if err != nil {
-			logger.Error(err, "failed to fetch KongIngress resource for Services",
-				"names", PrettyPrintServiceList(ks.Upstreams[i].Service.K8sServices))
-			continue
+			failuresCollector.PushResourceFailure(err.Error(), lo.Map(servicesGroup, servicesAsObjects)...)
+		} else {
+			for _, svc := range ks.Upstreams[i].Service.K8sServices {
+				ks.Upstreams[i].override(kongIngress, svc)
+			}
 		}
 
-		for _, svc := range ks.Upstreams[i].Service.K8sServices {
-			ks.Upstreams[i].override(kongIngress, svc)
+		kongUpstreamPolicy, err := GetKongUpstreamPolicyForServices(s, servicesGroup)
+		if err != nil {
+			failuresCollector.PushResourceFailure(err.Error(), lo.Map(servicesGroup, servicesAsObjects)...)
+		} else {
+			if kongUpstreamPolicy != nil {
+				if kongIngress != nil {
+					for _, svc := range servicesGroup {
+						logger.Error(nil,
+							fmt.Sprintf("Service uses both %s and %s annotations, should use only %s annotation. Settings "+
+								"from %s will take precedence",
+								annotations.AnnotationPrefix+annotations.ConfigurationKey,
+								annotations.AnnotationPrefix+kongv1beta1.KongUpstreamPolicyAnnotationKey,
+								annotations.AnnotationPrefix+kongv1beta1.KongUpstreamPolicyAnnotationKey,
+								annotations.AnnotationPrefix+kongv1beta1.KongUpstreamPolicyAnnotationKey),
+							"namespace", svc.Namespace, "name", svc.Name)
```

Comment:
Added 3d82093956152b34edfca8a8f2dbaca89456ce5e as hopefully not controversial and faster than opening a PR into this branch. This is a relatively benign misconfiguration, but users should use only one or the other.

Reply:
Let's handle https://github.com/Kong/kubernetes-ingress-controller/pull/5022 changes separately and do not merge them to this PR (it's already huge enough).